daemon/c_seccomp/sysinfo.c

```diff
@@ -0,0 +1,406 @@
+		      load[1], load[2], load[3], load[4], load[5]);
+
+		sysinfo_params->info->loads[0] =
+			(load[0] << SI_LOAD_SHIFT) + ((load[1] << SI_LOAD_SHIFT) / 100);
+		sysinfo_params->info->loads[1] =
+			(load[2] << SI_LOAD_SHIFT) + ((load[3] << SI_LOAD_SHIFT) / 100);
+		sysinfo_params->info->loads[2] =
+			(load[4] << SI_LOAD_SHIFT) + ((load[5] << SI_LOAD_SHIFT) / 100);
+	}
+	mem_free0(loadavg);
+
+	// overwrite totalram and freeram if memory limits are set
+	unsigned long mem_max = 0;
+	if (-1 == c_seccomp_cgroup_get_mem_max(&mem_max)) {
+		WARN("Failed to get memory.max from cgroup");
+	} else if (mem_max != MEM_UNLIMITED) {
+		// if cgroup is unlimited mem_max is set to 0 (MEM_UNLIMITED)
+		sysinfo_params->info->totalram = mem_max;
+
+		// overwrite freeram
+		unsigned long mem_current = 0;
+		if (-1 == c_seccomp_cgroup_get_mem_current(&mem_current)) {
+			WARN("Failed to get memory.current from cgroup");
+		} else {
+			sysinfo_params->info->freeram = mem_max - mem_current;
+		}
+	}
+
+	// overwrite sharedram
+	struct cg_mem_stat mem_stat = { 0 };
+	if (-1 == c_seccomp_cgroup_get_mem_stat(&mem_stat)) {
+		WARN("Failed to get mem_stat from cgroup");
+	} else {
+		sysinfo_params->info->sharedram = mem_stat.shmem;
+	}
+	// equivalent for bufferram does not exist in cgroups v2 memory.stat
+	sysinfo_params->info->bufferram = 0ULL;
+
+	// overwrite totalswap and freeswap if swap limits are set
+	unsigned long swap_max = 0;
+	if (-1 == c_seccomp_cgroup_get_swap_max(&swap_max)) {
+		WARN("Failed to get memory.swap.max from cgroup");
+	} else if (swap_max != MEM_UNLIMITED) {
+		// if cgroup is unlimited swap_max is set to 0 (MEM_UNLIMITED)
+		sysinfo_params->info->totalswap = swap_max;
+
+		// overwrite freeswap
+		unsigned long swap_current = 0;
+		if (-1 == c_seccomp_cgroup_get_swap_current(&swap_current)) {
+			WARN("Failed to get memory.swap.current from cgroup");
+		} else {
+			sysinfo_params->info->freeswap = swap_max - swap_current;
+		}
+	}
+
+	// overwrite procs
+	unsigned short pids_current = 0;
+	if (-1 == c_seccomp_cgroup_get_pids_current(&pids_current)) {
+		WARN("Failed to get pids.current from cgroup");
+	} else {
+		sysinfo_params->info->procs = pids_current;
+	}
+
+	TRACE("sysinfo struct emulated!");
+	c_seccomp_print_sysinfo(sysinfo_params->info);
+
+	if (-1 == c_seccomp_send_vm(sysinfo_params->seccomp, sysinfo_params->target_pid,
+				    (void *)sysinfo_params->info, sysinfo_params->target_datap,
+				    sizeof(struct sysinfo))) {
+		ERROR_ERRNO("Failed to send struct sysinfo");
+		return -1;
+	}
+
+	return 0;
+}
+
+int
+c_seccomp_emulate_sysinfo(c_seccomp_t *seccomp, struct seccomp_notif *req,
+			  struct seccomp_notif_resp *resp)
+{
+	int ret_sysinfo = 0;
+	struct sysinfo *info;
+
+	DEBUG("Got sysinfo, struct sysinfo *: %p", (void *)req->data.args[0]);
+	if (!(info = (struct sysinfo *)c_seccomp_fetch_vm_new(
+		      seccomp, req->pid, (void *)req->data.args[0], sizeof(struct sysinfo)))) {
```

Comment:
If sysinfo only writes data to the sysinfo struct, do we even have to fetch the struct first? Can't we just write to the pointer passed to the syscall? Or does this serve as a kind of pointer validation?

Reply:
Yes true. We could skip this here.